src/gtfs/mod.rs

```diff
@@ -341,6 +348,201 @@ pub struct Configuration {
+    fn try_from(attribution: &Attribution) -> Result<Self> {
+        let (object_type, object_id) = match (
+            attribution.route_id.is_some(),
+            attribution.trip_id.is_some(),
+        ) {
+            (true, false) => (Some(ObjectType::Route), attribution.route_id.clone()),
+            (false, true) => (
+                Some(ObjectType::VehicleJourney),
+                attribution.trip_id.clone(),
+            ),
+            _ => (None, None),
+        };
+
+        if object_type.is_none() {
+            let message = if let Some(attribution_id) = &attribution.attribution_id {
+                format!(
+                    "Attribution {} must have either route_id or trip_id",
+                    attribution_id
+                )
+            } else {
+                format!(
+                    "Attribution {:?} must have either route_id or trip_id",
+                    attribution
+                )
+            };
+            bail!(message);
+        }
+
+        let attribution_rule = AttributionRule {
+            id: calculate_hash(&attribution).to_string(),
+            object_type: object_type.expect("an error occured"),
+            object_id: object_id.expect("an error occured"),
+            organization_name: attribution.organization_name.clone(),
+            attribution_url: attribution.attribution_url.clone(),
+            attribution_email: attribution.attribution_email.clone(),
+            attribution_phone: attribution.attribution_phone.clone(),
+        };
+        Ok(attribution_rule)
+    }
+}
+
+fn read_attributions<H>(file_handler: &mut H, file_name: &str) -> Result<Vec<AttributionRule>>
+where
+    for<'a> &'a mut H: FileHandler,
+{
+    let (reader, path) = file_handler.get_file_if_exists(file_name)?;
+    let file_name = path.file_name();
+    let basename = file_name.map_or(path.to_string_lossy(), |b| b.to_string_lossy());
+    let mut attribution_rules = Vec::new();
+    if let Some(reader) = reader {
+        info!("Reading {}", basename);
+        let mut rdr = csv::ReaderBuilder::new()
+            .flexible(true)
+            .trim(csv::Trim::All)
+            .from_reader(reader);
+
+        for result in rdr.deserialize() {
+            let attribution: Attribution = result?;
+            if let Some(is_operator) = attribution.is_operator {
+                if is_operator {
+                    let attribution_rule =
+                        skip_error_and_warn!(AttributionRule::try_from(&attribution));
+                    attribution_rules.push(attribution_rule);
+                }
+            }
+        }
+    };
+
+    Ok(attribution_rules)
+}
+
+fn calculate_hash<T: Hash>(t: &T) -> u64 {
+    let mut s = DefaultHasher::new();
+    t.hash(&mut s);
+    s.finish()
+}
+
+fn get_or_create_company(
+    companies: &mut CollectionWithId<objects::Company>,
+    attribution_rule: &AttributionRule,
+) -> Result<String> {
+    if companies.get(&attribution_rule.id).is_none() {
```

Comment:
You should be able to use `contains_id`

```suggestion
    if !companies.contains_id(&attribution_rule.id) {
```